[Bug] FTBFS with libxfce4windowing 4.19.3: undefined reference to `xfw_window_get_id'

### Description

`xfw_(window|application)_get_id()` is removed in libxfce4windowing 4.19.3: https://gitlab.xfce.org/xfce/libxfce4windowing/-/commit/d7d9bae6294df6fb0b6cefa50d6fe8ee3d1ce800


This causes FTBFS as:

```
[287/478] Linking target src/daemon/budgie-daemon
FAILED: src/daemon/budgie-daemon 
gcc  -o src/daemon/budgie-daemon src/daemon/budgie-daemon.p/meson-generated_endsession.c.o src/daemon/budgie-daemon.p/meson-generated_notifications_dbus.c.o src/daemon/budgie-daemon.p/meson-generated_notifications_popup.c.o src/daemon/budgie-daemon.p/meson-generated_screenshot.c.o src/daemon/budgie-daemon.p/meson-generated_statusnotifier.c.o src/daemon/budgie-daemon.p/meson-generated_main.c.o src/daemon/budgie-daemon.p/meson-generated_manager.c.o src/daemon/budgie-daemon.p/meson-generated_menus.c.o src/daemon/budgie-daemon.p/meson-generated_osd.c.o src/daemon/budgie-daemon.p/meson-generated_settings.c.o src/daemon/budgie-daemon.p/meson-generated_tabswitcher.c.o src/daemon/budgie-daemon.p/meson-generated_xdgdirtracker_dbus.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group src/daemon/libbudgie-daemon-resources.a -Wl,--no-whole-archive '-Wl,-rpath,$ORIGIN/../lib:$ORIGIN/../plugin/panel:$ORIGIN/../theme' -Wl,-rpath-link,/build/source/build/src/lib -Wl,-rpath-link,/build/source/build/src/plugin/panel -Wl,-rpath-link,/build/source/build/src/theme src/lib/libbudgie-private.so.0.0.0 src/plugin/panel/libbudgie-plugin.so.0.0.0 src/appsys/libappsys.a src/config/libconfig.a src/theme/libbudgietheme.so.0.0.0 src/libsession/libsession.a /nix/store/rblppj260x50jd53dr1zgrs5q9aj4myx-libcanberra-0.30/lib/libcanberra.so /nix/store/rblppj260x50jd53dr1zgrs5q9aj4myx-libcanberra-0.30/lib/libcanberra-gtk3.so /nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib/libgtk-3.so /nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib/libgdk-3.so /nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib/libpangocairo-1.0.so /nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib/libpango-1.0.so /nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib/libharfbuzz.so /nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib/libatk-1.0.so /nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib/libcairo-gobject.so /nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib/libcairo.so /nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib/libgdk_pixbuf-2.0.so /nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib/libgio-2.0.so /nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib/libgobject-2.0.so /nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib/libglib-2.0.so -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -L/nix/store/rblppj260x50jd53dr1zgrs5q9aj4myx-libcanberra-0.30/lib -lX11 -lz -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -lz -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -lz /nix/store/2d3x7vlpgxmg68aq6zsaspz12gh8nc7j-libnotify-0.8.3/lib/libnotify.so /nix/store/zvilx2vl50a2dmabfrraw6fsd69y5wdv-libwnck-43.0/lib/libwnck-3.so -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -L/nix/store/zvilx2vl50a2dmabfrraw6fsd69y5wdv-libwnck-43.0/lib -lz /nix/store/9rjhwfasx8wbfpb7j5c908c20dx1ria2-libxfce4windowing-4.19.3/lib/libxfce4windowing-0.so -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -L/nix/store/9rjhwfasx8wbfpb7j5c908c20dx1ria2-libxfce4windowing-4.19.3/lib -lz /nix/store/9yx2a6qmpg42mp0nqr6qh75imdqdbfwk-gstreamer-1.22.9/lib/libgstreamer-1.0.so /nix/store/nh3w2ckfcwi9lvysyrg2l394a4ckw2gi-libpeas-1.36.0/lib/libpeas-1.0.so -Wl,--export-dynamic /nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib/libgmodule-2.0.so -pthread /nix/store/bg3qymmzpfmv8n0bj8kbv3ha6nas85hf-gobject-introspection-1.78.1/lib/libgirepository-1.0.so -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -lz -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -lz -L/nix/store/lk8kgif9ynk1l6aiqnbvzvys83n0h956-gtk+3-3.24.41/lib -L/nix/store/gqf5ixhbfsizc15ndmdnir8bir5b87qj-at-spi2-core-2.50.0/lib -L/nix/store/g0gc1ls4p1rwcx7k5qr153rs71ir8ixn-glib-2.78.4/lib -L/nix/store/jp83hbx84zxiqmpprcj8g5ff0jlmh7rz-cairo-1.18.0/lib -L/nix/store/jrlkyy3pzizi39lywafsd1psd9vg3p9b-gdk-pixbuf-2.42.10/lib -L/nix/store/ahf3r66zp3c8hg0y9b151xr29x8b6gjv-pango-1.51.0/lib -L/nix/store/c7m7x38l8xlxlgqfgca0753pzzhki90r-harfbuzz-8.3.0/lib -L/nix/store/zvilx2vl50a2dmabfrraw6fsd69y5wdv-libwnck-43.0/lib -lz -Wl,--end-group
/nix/store/p0p56gzz837fgmfd7lyisghcd7x2fdlc-binutils-2.40/bin/ld: src/daemon/budgie-daemon.p/meson-generated_tabswitcher.c.o: in function `_budgie_tab_switcher_window_set_window_as_activated_budgie_tab_switcher_widget_window_activated':
tabswitcher.c:(.text+0x71f): undefined reference to `xfw_window_get_id'
/nix/store/p0p56gzz837fgmfd7lyisghcd7x2fdlc-binutils-2.40/bin/ld: src/daemon/budgie-daemon.p/meson-generated_tabswitcher.c.o: in function `budgie_tab_switcher_widget_construct':
tabswitcher.c:(.text+0xc50): undefined reference to `xfw_window_get_id'
/nix/store/p0p56gzz837fgmfd7lyisghcd7x2fdlc-binutils-2.40/bin/ld: src/daemon/budgie-daemon.p/meson-generated_tabswitcher.c.o: in function `budgie_tab_switcher_window_focus_item':
tabswitcher.c:(.text+0x2062): undefined reference to `xfw_window_get_id'
collect2: error: ld returned 1 exit status
```

Full log: https://gist.github.com/bobby285271/6ad799ceae415f6f32a472eaac06cbe1


### Budgie version

10.9.1

### Operating System

NixOS (using custom checkout)

### Steps to reproduce the issue

1. Build budgie-desktop with libxfce4windowing 4.19.3

### Actual result

FTBFS.

### Expected result

Builds.

### Additional information

_No response_

--- a/src/daemon/tabswitcher.vala
+++ b/src/daemon/tabswitcher.vala
@@ -61,7 +61,7 @@ namespace Budgie {
 		public TabSwitcherWidget(Budgie.AppSystem app_system, libxfce4windowing.Window? win) {
 			Object();
 			window = win;
-			var uid = window.get_id();
+			ulong uid = (ulong) window.x11_get_xid();
 			id = uid.to_string();
 			set_title();
 
@@ -334,7 +334,7 @@ namespace Budgie {
 		}
 
 		private void set_window_as_activated(libxfce4windowing.Window window) {
-			string id = window.get_id().to_string();
+			string id = ((ulong) window.x11_get_xid()).to_string();
 			unowned List<string> entries = recency.find_custom(id, strcmp);
 			recency.remove_link(entries);
 			recency.prepend(id);
@@ -365,7 +365,7 @@ namespace Budgie {
 		/* Switch focus to the item with the xid */
 		public void focus_item(bool backwards) {
 			unowned libxfce4windowing.Window? active_window = xfce_screen.get_active_window();
-			TabSwitcherWidget? widget = active_window != null ? ids.get(active_window.get_id().to_string()) : null;
+			TabSwitcherWidget? widget = active_window != null ? ids.get(((ulong) active_window.x11_get_xid()).to_string()) : null;
 
 			// Visible, each input should cycle to previous / next
 			if (visible) {

--- a/src/panel/applets/workspaces/WindowIcon.vala
+++ b/src/panel/applets/workspaces/WindowIcon.vala
@@ -72,7 +72,7 @@ namespace Workspaces {
 		}
 
 		public void on_drag_data_get(Gtk.Widget widget, Gdk.DragContext context, Gtk.SelectionData selection_data, uint target_type, uint time) {
-			ulong window_xid = (ulong)window.get_id();
+			ulong window_xid = (ulong)window.x11_get_xid();
 			uchar[] buf;
 			convert_ulong_to_bytes(window_xid, out buf);
 			selection_data.set(

--- a/src/panel/applets/workspaces/WorkspaceItem.vala
+++ b/src/panel/applets/workspaces/WorkspaceItem.vala
@@ -179,7 +179,7 @@ namespace Workspaces {
 			if (data != null) {
 				try {
 					foreach (libxfce4windowing.Window window in WorkspacesApplet.xfce_screen.get_windows()) {
-						if (window.get_id() == *data) {
+						if (window.x11_get_xid() == *data) {
 							window.move_to_workspace(this.workspace);
 							dnd_success = true;
 							break;

--- a/src/panel/applets/workspaces/WorkspacesApplet.vala
+++ b/src/panel/applets/workspaces/WorkspacesApplet.vala
@@ -278,8 +278,8 @@ namespace Workspaces {
 		}
 
 		private void connect_signals() {
-			connections += workspace_group.workspace_created.connect(workspace_added);
-			connections += workspace_group.workspace_destroyed.connect(workspace_removed);
+			connections += workspace_group.workspace_added.connect(workspace_added);
+			connections += workspace_group.workspace_removed.connect(workspace_removed);
 			connections += workspace_group.active_workspace_changed.connect(set_current_workspace);
 			connections += xfce_screen.active_window_changed.connect(update_workspaces);
 			connections += xfce_screen.window_opened.connect(window_opened);
@@ -403,7 +403,7 @@ namespace Workspaces {
 
 			libxfce4windowing.Window? window = null;
 			foreach (libxfce4windowing.Window win in xfce_screen.get_windows()) {
-				if (win.get_id() == *data) {
+				if (win.x11_get_xid() == *data) {
 					window = win;
 					break;
 				}

--- a/vapi/README.md
+++ b/vapi/README.md
@@ -1,17 +1,36 @@
-Refreshing vapis
---------------
+## Refreshing vapis
+
+### Polkit
 
 To refresh the Polkit vapi files:
 
-    vapigen --library polkit-gobject-1 /usr/share/gir-1.0/Polkit-1.0.gir --pkg gio-unix-2.0
-    vapigen --library polkit-agent-1 /usr/share/gir-1.0/PolkitAgent-1.0.gir --pkg gio-unix-2.0 --pkg polkit-gobject-1 --girdir=. --vapidir=.
+```bash
+vapigen --library polkit-gobject-1 /usr/share/gir-1.0/Polkit-1.0.gir --pkg gio-unix-2.0
+vapigen --library polkit-agent-1 /usr/share/gir-1.0/PolkitAgent-1.0.gir --pkg gio-unix-2.0 --pkg polkit-gobject-1 --girdir=. --vapidir=.
+```
 
-Then have fun un-mangling it to support vala async syntax
+Then have fun un-mangling it to support vala async syntax.
+
+### Libxfce4windowing
+
+To generate the libxfce4windowing vapi files:
+
+```bash
+vapigen --library libxfce4windowing-0 /usr/share/gir-1.0/Libxfce4windowing-0-0.gir --metadatadir .
+```
+
+### UPower
 
 To generate the UPower vapi files:
 
-    vapigen --library upower-glib /usr/share/gir-1.0/UpowerGlib-1.0.gir --metadatadir . --pkg gio-unix-2.0 UPowerGlib-1.0-custom.vala
+```bash
+vapigen --library upower-glib /usr/share/gir-1.0/UpowerGlib-1.0.gir --metadatadir . --pkg gio-unix-2.0 UPowerGlib-1.0-custom.vala
+```
+
+### Mutter
 
 For mutter (and shipped cogl and clutter), once you defined the relative `*.deps`, `*.metadata` and `*-custom.vala` files, you can run:
 
-    ./vapi/generate-mutter-vapi.sh <mutter-version>
+```bash
+./vapi/generate-mutter-vapi.sh <mutter-version>
+```

--- a/vapi/libxfce4windowing-0.deps
+++ b/vapi/libxfce4windowing-0.deps
@@ -1,3 +1,4 @@
 gdk-3.0
 glib-2.0
+gtk+-3.0
 libwnck-3.0

--- a/vapi/libxfce4windowing-0.vapi
+++ b/vapi/libxfce4windowing-0.vapi
@@ -6,13 +6,14 @@ namespace libxfce4windowing {
 	public abstract class Application : GLib.Object {
 		[CCode (has_construct_function = false)]
 		protected Application ();
+		[CCode (cname = "xfw_application_get_class_id")]
+		[Version (since = "4.19.3")]
+		public unowned string get_class_id ();
 		[CCode (cname = "xfw_application_get_gicon")]
 		[Version (since = "4.19.1")]
 		public unowned GLib.Icon get_gicon ();
 		[CCode (cname = "xfw_application_get_icon")]
 		public unowned Gdk.Pixbuf? get_icon (int size, int scale);
-		[CCode (cname = "xfw_application_get_id")]
-		public uint64 get_id ();
 		[CCode (cname = "xfw_application_get_instance")]
 		public unowned libxfce4windowing.ApplicationInstance? get_instance (libxfce4windowing.Window window);
 		[CCode (cname = "xfw_application_get_instances")]
@@ -25,9 +26,10 @@ namespace libxfce4windowing {
 		[Version (since = "4.19.1")]
 		public bool icon_is_fallback ();
 		[NoAccessorMethod]
-		public GLib.Icon gicon { owned get; }
+		[Version (since = "4.19.3")]
+		public string class_id { owned get; }
 		[NoAccessorMethod]
-		public uint64 id { get; }
+		public GLib.Icon gicon { owned get; }
 		[NoAccessorMethod]
 		public void* instances { get; }
 		[NoAccessorMethod]
@@ -87,6 +89,9 @@ namespace libxfce4windowing {
 		public unowned libxfce4windowing.Application get_application ();
 		[CCode (cname = "xfw_window_get_capabilities")]
 		public libxfce4windowing.WindowCapabilities get_capabilities ();
+		[CCode (array_length = false, array_null_terminated = true, cname = "xfw_window_get_class_ids")]
+		[Version (since = "4.19.3")]
+		public unowned string[] get_class_ids ();
 		[CCode (cname = "xfw_window_get_geometry")]
 		public unowned Gdk.Rectangle? get_geometry ();
 		[CCode (cname = "xfw_window_get_gicon")]
@@ -94,8 +99,6 @@ namespace libxfce4windowing {
 		public unowned GLib.Icon get_gicon ();
 		[CCode (cname = "xfw_window_get_icon")]
 		public unowned Gdk.Pixbuf? get_icon (int size, int scale);
-		[CCode (cname = "xfw_window_get_id")]
-		public uint64 get_id ();
 		[CCode (cname = "xfw_window_get_monitors")]
 		public unowned GLib.List<Gdk.Monitor>? get_monitors ();
 		[CCode (cname = "xfw_window_get_name")]
@@ -165,14 +168,19 @@ namespace libxfce4windowing {
 		public bool start_move () throws GLib.Error;
 		[CCode (cname = "xfw_window_start_resize")]
 		public bool start_resize () throws GLib.Error;
+		[CCode (cname = "xfw_window_x11_get_xid")]
+		[Version (since = "4.19.3")]
+		public X.Window x11_get_xid ();
 		[NoAccessorMethod]
 		public libxfce4windowing.Application application { owned get; }
 		[NoAccessorMethod]
 		public libxfce4windowing.WindowCapabilities capabilities { get; }
+		[CCode (array_length = false, array_null_terminated = true)]
+		[NoAccessorMethod]
+		[Version (since = "4.19.3")]
+		public string[] class_ids { owned get; }
 		[NoAccessorMethod]
 		public GLib.Icon gicon { owned get; }
-		[NoAccessorMethod]
-		public uint64 id { get; }
 		[NoAccessorMethod]
 		public void* monitors { get; }
 		[NoAccessorMethod]
@@ -186,6 +194,8 @@ namespace libxfce4windowing {
 		[NoAccessorMethod]
 		public libxfce4windowing.Workspace workspace { owned get; }
 		public signal void capabilities_changed (libxfce4windowing.WindowCapabilities changed_mask, libxfce4windowing.WindowCapabilities new_state);
+		[Version (since = "4.19.3")]
+		public signal void class_changed ();
 		public signal void closed ();
 		public signal void geometry_changed ();
 		public signal void icon_changed ();
@@ -313,6 +323,8 @@ namespace libxfce4windowing {
 	public interface Workspace : GLib.Object {
 		[CCode (cname = "xfw_workspace_activate")]
 		public bool activate () throws GLib.Error;
+		[CCode (cname = "xfw_workspace_assign_to_workspace_group")]
+		public bool assign_to_workspace_group (libxfce4windowing.WorkspaceGroup group) throws GLib.Error;
 		[CCode (cname = "xfw_workspace_get_capabilities")]
 		public libxfce4windowing.WorkspaceCapabilities get_capabilities ();
 		[CCode (cname = "xfw_workspace_get_geometry")]
@@ -332,13 +344,13 @@ namespace libxfce4windowing {
 		[CCode (cname = "xfw_workspace_get_state")]
 		public libxfce4windowing.WorkspaceState get_state ();
 		[CCode (cname = "xfw_workspace_get_workspace_group")]
-		public unowned libxfce4windowing.WorkspaceGroup get_workspace_group ();
+		public unowned libxfce4windowing.WorkspaceGroup? get_workspace_group ();
 		[CCode (cname = "xfw_workspace_remove")]
 		public bool remove () throws GLib.Error;
 		[NoAccessorMethod]
 		public abstract libxfce4windowing.WorkspaceCapabilities capabilities { get; set construct; }
 		[NoAccessorMethod]
-		public abstract libxfce4windowing.WorkspaceGroup group { owned get; construct; }
+		public abstract libxfce4windowing.WorkspaceGroup group { owned get; }
 		[NoAccessorMethod]
 		public abstract string id { owned get; }
 		[NoAccessorMethod]
@@ -348,6 +360,7 @@ namespace libxfce4windowing {
 		[NoAccessorMethod]
 		public abstract libxfce4windowing.WorkspaceState state { get; set construct; }
 		public signal void capabilities_changed (libxfce4windowing.WorkspaceCapabilities changed_mask, libxfce4windowing.WorkspaceCapabilities new_capabilities);
+		public signal void group_changed (libxfce4windowing.WorkspaceGroup previous_group);
 		public signal void name_changed ();
 		public signal void state_changed (libxfce4windowing.WorkspaceState changed_mask, libxfce4windowing.WorkspaceState new_state);
 	}
@@ -383,19 +396,31 @@ namespace libxfce4windowing {
 		public abstract void* workspaces { get; }
 		public signal void active_workspace_changed (libxfce4windowing.Workspace? previously_active_workspace);
 		public signal void capabilities_changed (libxfce4windowing.WorkspaceGroupCapabilities changed_mask, libxfce4windowing.WorkspaceGroupCapabilities new_capabilities);
+		public signal void monitor_added (Gdk.Monitor monitor);
+		public signal void monitor_removed (Gdk.Monitor monitor);
 		public signal void monitors_changed ();
 		public signal void viewports_changed ();
-		public signal void workspace_created (libxfce4windowing.Workspace workspace);
-		public signal void workspace_destroyed (libxfce4windowing.Workspace workspace);
+		public signal void workspace_added (libxfce4windowing.Workspace workspace);
+		public signal void workspace_removed (libxfce4windowing.Workspace workspace);
 	}
 	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "XfwWorkspaceManager", type_id = "xfw_workspace_manager_get_type ()")]
 	public interface WorkspaceManager : GLib.Object {
 		[CCode (cname = "xfw_workspace_manager_list_workspace_groups")]
 		public unowned GLib.List<libxfce4windowing.WorkspaceGroup>? list_workspace_groups ();
+		[CCode (cname = "xfw_workspace_manager_list_workspaces")]
+		public unowned GLib.List<libxfce4windowing.Workspace>? list_workspaces ();
 		[NoAccessorMethod]
 		public abstract Gdk.Screen screen { owned get; construct; }
+		public signal void workspace_created (libxfce4windowing.Workspace workspace);
+		public signal void workspace_destroyed (libxfce4windowing.Workspace workspace);
 		public signal void workspace_group_created (libxfce4windowing.WorkspaceGroup group);
 		public signal void workspace_group_destroyed (libxfce4windowing.WorkspaceGroup group);
+	}
+	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "XfwClientType", cprefix = "XFW_CLIENT_TYPE_", has_type_id = false)]
+	[Version (since = "4.19.3")]
+	public enum ClientType {
+		APPLICATION,
+		PAGER
 	}
 	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "XfwDirection", cprefix = "XFW_DIRECTION_", type_id = "xfw_direction_get_type ()")]
 	public enum Direction {
@@ -507,6 +532,9 @@ namespace libxfce4windowing {
 	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "xfw_screen_get_default")]
 	[Version (replacement = "Screen.get_default")]
 	public static libxfce4windowing.Screen screen_get_default ();
+	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "xfw_set_client_type")]
+	[Version (since = "4.19.3")]
+	public static void set_client_type (libxfce4windowing.ClientType client_type);
 	[CCode (cheader_filename = "libxfce4windowing/libxfce4windowing.h", cname = "xfw_windowing_get")]
 	public static libxfce4windowing.Windowing windowing_get ();
 }

